test/test_optimizers.f90

```diff
@@ -0,0 +1,150 @@
+
+    call net(1) % forward(x)
+    call net(1) % backward(y)
+    call net(1) % update(optimizer=sgd(learning_rate=1.))
+
+    if (mod(n, 10) == 0) then
+      ypred = net(1) % predict(x)
+      converged = check_convergence(ypred, y, prev_loss)
+      if (converged) then
+        exit
+      end if
+      prev_loss = loss
+    end if
+
+  end do
+
+  if (.not. converged) then
+    write(stderr, '(a)') &
+      'sgd should converge in simple training.. failed'
+    ok_sgd = .false.
+  end if
+
+  ! Resetting loss and convergence flag
+  converged = .false.
+  prev_loss = 0.0
+
+  do n = 0, num_iterations
+
+    call net(2) % forward(x)
+    call net(2) % backward(y)
+    call net(2) % update(optimizer=sgd(learning_rate=1., momentum=0.9))
+
+    if (mod(n, 10) == 0) then
+      ypred = net(2) % predict(x)
+      converged = check_convergence(ypred, y, prev_loss)
+      if (converged) then
+        exit
+      end if
+      prev_loss = loss
+    end if
+
+  end do
+
+  if (.not. converged) then
+    write(stderr, '(a)') &
+      'sgd(momentum) should converge in simple training.. failed'
+    ok_momentum = .false.
+  end if
+
+  ! Resetting loss and convergence flag
+  converged = .false.
+  prev_loss = 0.0
+
+  do n = 0, num_iterations
+
+    call net(3) % forward(x)
+    call net(3) % backward(y)
+    call net(3) % update(optimizer=sgd(learning_rate=1., momentum=0.9, nesterov=.true.))
+
+    if (mod(n, 10) == 0) then
+      ypred = net(3) % predict(x)
+      converged = check_convergence(ypred, y, prev_loss)
+      if (converged) then
+        exit
+      end if
+      prev_loss = loss
+    end if
+
+  end do
+
+  if (.not. converged) then
+    write(stderr, '(a)') &
+      'sgd(nesterov) should converge in simple training.. failed'
+    ok_nesterov = .false.
+  end if
+
+  ! Resetting loss and convergence flag
+  converged = .false.
+  prev_loss = 0.0
+
+  do n = 0, num_iterations
+
+    call net(4) % forward(x)
+    call net(4) % backward(y)
+    call net(4) % update(optimizer=rmsprop(learning_rate=0.01, decay_rate=0.9))
+
+    if (mod(n, 10) == 0) then
+      ypred = net(4) % predict(x)
+      converged = check_convergence(ypred, y, prev_loss)
+      if (converged) then
+        exit
+      end if
+      prev_loss = loss
+    end if
+
+  end do
+
+  if (.not. converged) then
+    write(stderr, '(a)') &
+    'rmsprop should converge in simple training.. failed'
+    ok_rmsprop = .false.
+  end if
+
+  if (ok_sgd .and. ok_momentum .and. ok_nesterov .and. ok_rmsprop) then
+    print '(a)', 'test_optimizers: All tests passed.'
+  else
+    write(stderr, '(a)') 'test_optimizers: One or more tests failed.'
+    stop 1
+  end if
+
+  contains
+
+ logical function check_convergence(ypred, y, prev_loss) result(converged)
+    real, intent(in) :: ypred(:), y(:), prev_loss
+
+    loss = sqrt(sum((ypred - y)**2) / size(y))
+    ! Check convergence.
+    if (abs(loss - prev_loss) < tolerance) then
+      converged = .true.
+    end if
```

Comment:
We actually need to test for convergence of loss to some target low value, not to the value of loss from the previous iteration. Currently these convergence tests are passing but if you print out the network outputs you'll see that the solution is not actually converging because you're exiting the loop too early. Rather checking for `abs(loss - prev_loss) < tolerance`, you need to check for `loss < tolerance`.

(I checked and the solutions are actually converging which is good news; you just need to fix the convergence tests).

Reply:
This should be a pure function.

Reply:
Notice also that if the convergence test doesn't pass, you don't actually set the function result to `.false.`. A better way to write this is:

```fortran
converged = loss < tolerance
```